src/mavsdk/plugins/component_metadata/file_cache.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <filesystem>
+#include <cstdint>
+#include <optional>
+#include <string>
+#include <map>
+
+namespace mavsdk {
+
+/**
+ * Simple file cache with a maximum number of files and LRU retention policy based on last access
+ * Notes:
+ * - fileTag defines the cache keys and the format is up to the user
+ * - thread-safe
+ * - multiple instances (or processes) can work on the same directory.
+ *   (however there is no lock held when returning a file from a cache, so the assumption is that
+ *   the cache is large enough so no other process evicts the file until it is used).
+ */
+class FileCache {
+public:
+    FileCache(std::filesystem::path path, int max_num_files);
+
+    /**
+     * Try to access a file and set the access counter
+     * @param fileTag
+     * @return file path or nullopt if not found
+     */
+    std::optional<std::filesystem::path> access(const std::string& file_tag);
+
+    /**
+     * Insert a file into the cache & remove old files if there's too many.
+     * @param fileTag
+     * @param fileName file to insert, will be moved (or deleted if already exists)
+     * @return cached file name if inserted or already exists, nullopt on error
+     */
+    std::optional<std::filesystem::path>
+    insert(const std::string& file_tag, const std::filesystem::path& file_name);
+
+private:
+    class DirectoryLock {
+    public:
+        explicit DirectoryLock(const std::filesystem::path& path);
+        ~DirectoryLock();
+
+    private:
+#ifdef _WIN32
+        OVERLAPPED _overlapped{};
+        const int _lock_size = 10000;
+        HANDLE _lock{};
+#else
+        int _fd{-1};
+#endif
+        const std::filesystem::path _path;
+    };
+
+    static constexpr const char* _meta_extension = ".meta";
+    static constexpr const char* _cache_extension = ".cache";
+    static constexpr const char* _lock_file = "lock";
+
+    using AccessCounterType = uint64_t;
+
+    struct Meta {
+        uint32_t magic{0x9a9cad0e};
+        uint32_t version{0};
+        AccessCounterType access_counter{0};
+    };
+
+    struct AccessCounters {
+        AccessCounterType next_access_counter{0};
+        std::map<AccessCounterType, std::string> cached_files;
```

Comment:
Looks like it can be `std::unordered_map`.

Reply:
No, the cache eviction depends on the order (least recently used)

Reply:
Oh, I see.